File: src/Activities/Turbo.Activities.Hiking.Contracts/value/HikingDetails.cs
using System.Text.Json.Serialization;

namespace Turboapi.Activities.Hiking.value;

public sealed record HikingDetails
{
    [JsonPropertyName("distanceMeters")] public int DistanceMeters { get; init; }
    [JsonPropertyName("ascentMeters")] public int AscentMeters { get; init; }
    [JsonPropertyName("descentMeters")] public int DescentMeters { get; init; }
    [JsonPropertyName("elevationMinMeters")] public int ElevationMinMeters { get; init; }
    [JsonPropertyName("elevationMaxMeters")] public int ElevationMaxMeters { get; init; }
    [JsonPropertyName("difficulty")] public HikingDifficulty Difficulty { get; init; }
    [JsonPropertyName("surface")] public TrailSurface Surface { get; init; }
    [JsonPropertyName("marking")] public TrailMarking Marking { get; init; }
    [JsonPropertyName("estimatedHours")] public float? EstimatedHours { get; init; }
    [JsonPropertyName("hasWaterSources")] public bool HasWaterSources { get; init; }
    [JsonPropertyName("hasShelter")] public bool HasShelter { get; init; }
    [JsonPropertyName("waterSources")] public IReadOnlyList<WaterSource> WaterSources { get; init; } = Array.Empty<WaterSource>();

    [JsonConstructor]
    public HikingDetails(
        int distanceMeters, int ascentMeters, int descentMeters,
        int elevationMinMeters, int elevationMaxMeters,
        HikingDifficulty difficulty, TrailSurface surface, TrailMarking marking,
        float? estimatedHours, bool hasWaterSources, bool hasShelter,
        IReadOnlyList<WaterSource>? waterSources)
    {
        DistanceMeters = distanceMeters;
        AscentMeters = ascentMeters;
        DescentMeters = descentMeters;
        ElevationMinMeters = elevationMinMeters;
        ElevationMaxMeters = elevationMaxMeters;
        Difficulty = difficulty;
        Surface = surface;
        Marking = marking;
        EstimatedHours = estimatedHours;
        HasWaterSources = hasWaterSources;
        HasShelter = hasShelter;
        WaterSources = waterSources ?? Array.Empty<WaterSource>();
    }
}

public sealed record WaterSource
{
    [JsonPropertyName("lat")] public double Lat { get; init; }
    [JsonPropertyName("lon")] public double Lon { get; init; }
    [JsonPropertyName("kind")] public string Kind { get; init; }
    [JsonPropertyName("notes")] public string? Notes { get; init; }

    [JsonConstructor]
    public WaterSource(double lat, double lon, string kind, string? notes)
    {
        Lat = lat; Lon = lon;
        Kind = kind ?? throw new ArgumentNullException(nameof(kind));
        Notes = notes;
    }
}

File: lib/features/activity_hiking/models/hiking_details.dart
import 'package:latlong2/latlong.dart';

enum HikingDifficulty { easy, moderate, hard, expert }
enum TrailMarking { unmarked, cairns, paint, signposted }
enum TrailSurface { mixed, path, gravel, boardwalk, rock, scree, snow }

class HikingDetails {
  final int distanceMeters;
  final int ascentMeters;
  final int descentMeters;
  final int elevationMinMeters;
  final int elevationMaxMeters;
  final HikingDifficulty difficulty;
  final TrailSurface surface;
  final TrailMarking marking;
  final double? estimatedHours;
  final bool hasWaterSources;
  final bool hasShelter;
  final List<WaterSource> waterSources;

  const HikingDetails({
    required this.distanceMeters,
    required this.ascentMeters,
    required this.descentMeters,
    required this.elevationMinMeters,
    required this.elevationMaxMeters,
    required this.difficulty,
    required this.surface,
    required this.marking,
    this.estimatedHours,
    this.hasWaterSources = false,
    this.hasShelter = false,
    this.waterSources = const [],
  });

  Map<String, dynamic> toJson() => {
        'distanceMeters': distanceMeters,
        'ascentMeters': ascentMeters,
        'descentMeters': descentMeters,
        'elevationMinMeters': elevationMinMeters,
        'elevationMaxMeters': elevationMaxMeters,
        'difficulty': difficulty.index,
        'surface': surface.index,
        'marking': marking.index,
        'estimatedHours': ?estimatedHours,
        'hasWaterSources': hasWaterSources,
        'hasShelter': hasShelter,
        'waterSources': waterSources.map((w) => w.toJson()).toList(),
      };

  factory HikingDetails.fromJson(Map<String, dynamic> json) => HikingDetails(
        distanceMeters: (json['distanceMeters'] as num).toInt(),
        ascentMeters: (json['ascentMeters'] as num).toInt(),
        descentMeters: (json['descentMeters'] as num).toInt(),
        elevationMinMeters: (json['elevationMinMeters'] as num).toInt(),
        elevationMaxMeters: (json['elevationMaxMeters'] as num).toInt(),
        difficulty: HikingDifficulty.values[(json['difficulty'] as num).toInt()],
        surface: TrailSurface.values[(json['surface'] as num).toInt()],
        marking: TrailMarking.values[(json['marking'] as num).toInt()],
        estimatedHours: (json['estimatedHours'] as num?)?.toDouble(),
        hasWaterSources: json['hasWaterSources'] as bool? ?? false,
        hasShelter: json['hasShelter'] as bool? ?? false,
        waterSources: (json['waterSources'] as List?)
                ?.cast<Map<String, dynamic>>()
                .map(WaterSource.fromJson)
                .toList() ??
            const [],
      );
}

class WaterSource {
  final LatLng position;
  final String kind;
  final String? notes;
  const WaterSource({required this.position, required this.kind, this.notes});

  Map<String, dynamic> toJson() => {
        'lat': position.latitude,
        'lon': position.longitude,
        'kind': kind,
        'notes': ?notes,
      };

  factory WaterSource.fromJson(Map<String, dynamic> json) => WaterSource(
        position: LatLng(
          (json['lat'] as num).toDouble(),
          (json['lon'] as num).toDouble(),
        ),
        kind: json['kind'] as String,
        notes: json['notes'] as String?,
      );
}
